### PR TITLE
Support for orEmpty

### LINKED ISSUE #498: Support for orEmpty
Support for orEmpty

### PR DESCRIPTION
**Pull request by angelozerr**
Support for orEmpty

Fixes #498

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
Please don't test, there are so many bugs for the moment.

**Comment by angelozerr** [comment]
@fbricon It should work now, see the following demo:

OrEmptyValueResolverDemo

I need to clean some piece of code. Tests are written.

**Comment by angelozerr** [comment]
CI Build is failling because I need to adapt some tests but last push supports generic (for iterable, it doesn't work for Map for instance), see follwowing demo:

OrEmptyValueResolver2Demo

Please note completion list items doesn't resolve generic  (you see `get(int : index) : E` instead of `get(int : index) : Item`)for performance reason, we can see that in an another issue. 

The resolve of generic is working for hover, diagnostics and completion.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java [hunk 1]
```diff
@@ -145,7 +147,7 @@ private synchronized boolean computeIsIterable() {
 		if (iterableOf != null) {
 			return true;
 		}
-		boolean iterable = getSignature().equals("java.lang.Iterable");
+		boolean iterable = getName().equals("java.lang.Iterable");
```
Comment (AlexXuChen) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java:150]: We could use/define a utility class containing Java type constants

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java [hunk 2]
```diff
@@ -247,4 +249,14 @@ private static boolean isEmpty(String value) {
 		return value == null || value.isEmpty();
 	}
 
+	@Override
+	public String toString() {
+		ToStringBuilder b = new ToStringBuilder(this);
+		b.add("name", this.getName());
```
Comment (AlexXuChen) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java:255]: Also could define utility class/constants for Java type info

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java [hunk 3]
```diff
@@ -101,12 +114,35 @@ public String getSignature() {
...
+			return null;
+		}
+		return resolveReturnType(argType, parameter.getJavaType());
+	}
+
 	@Override
 	public String toString() {
 		ToStringBuilder b = new ToStringBuilder(this);
-		b.add("namespace", this.namespace);
+		b.add("namespace", this.getNamespace());
```
Comment (AlexXuChen) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java:140]: Same as above: constants for Java type info

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/JavaDataModelCache.java [hunk 4]
```diff
@@ -61,8 +62,74 @@ public JavaDataModelCache(QuteProjectRegistry projectRegistry) {
...
+		if (javaType.getExtendedTypes() != null) {
+			for (String extendedType : javaType.getExtendedTypes()) {
+				if (parameterType.equals(extendedType)) {
+					return true;
+				}
+			}
+		}
+		if (!javaType.getTypeParameters().isEmpty()) {
+			ResolvedJavaTypeInfo result = resolveJavaType(resolvedTypeName, projectUri).getNow(null);
+			if (result != null) {
```
Comment (AlexXuChen) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/JavaDataModelCache.java:122]: I believe Java supports short circuiting, so we can do `if (result != null && result.getExtendedTypes() != null) {`

**Review by AlexXuChen — CHANGES_REQUESTED** [review]
Feature is working as intended from my side, just a few small changes:

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java [hunk 5]
```diff
@@ -69,18 +58,42 @@ public String getSourceType() {
...
 	 * Set the source type.
-	 * 
+	 *
 	 * @param sourceType the source type.
 	 */
 	public void setSourceType(String sourceType) {
 		this.sourceType = sourceType;
 	}
 
+	public String getDescription() {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java:73]: fixed

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/ValueResolver.java [hunk 6]
```diff
@@ -69,18 +58,42 @@ public String getSourceType() {
...
 	 * Set the source type.
-	 * 
+	 *
 	 * @param sourceType the source type.
 	 */
 	public void setSourceType(String sourceType) {
 		this.sourceType = sourceType;
 	}
 
+	public String getDescription() {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/ValueResolver.java:73]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java (hunk 2 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java:259]: fixed,  but I have not created a constant class for that, we will see of we will need it in the future.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/JavaDataModelCache.java (hunk 4 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/project/datamodel/JavaDataModelCache.java]: good catch, fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java (hunk 1 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ResolvedJavaTypeInfo.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java (hunk 3 above)
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/commons/ValueResolver.java:170]: You mean set "namespace" as constants. Let's keep like this, the toString is just called when you debug the instance.

File: qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/JavaMemberInfo.java [hunk 7]
```diff
@@ -50,4 +50,8 @@ public String getSourceType() {
 		return resolvedType != null ? resolvedType.getSignature() : null;
 	}
 
+	public String resolveJavaElementType(ResolvedJavaTypeInfo resolvedJavaType) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/commons/JavaMemberInfo.java]: fixed

**Review by AlexXuChen — APPROVED** [review]
LGTM

**Pull request merged**